### ISSUE TITLE
SING_STEP seems not to be working for PDP-8/I

### ISSUE THREAD

**Issue by Thomas-Ganter** [simh/simh#773]
- #### Context

I completely fail to make SING_STEP work. I was told this is _by design_, however I find no such information in documentation (but also no paragraph documenting SING_STEP) or FAQ. 

Either my search term is wrong, or there is at minimum a documentation gap. 

- #### PDP-8 simulator V4.0-0 Current

```
sim> show version
PDP-8 simulator V4.0-0 Current
    Simulator Framework Capabilities:
        32b data
        32b addresses
        no Ethernet
        Idle/Throttling support is available
        Virtual Hard Disk (VHD) support
        RAW disk and CD/DVD ROM support
        Asynchronous I/O support (Lock free asynchronous event queue)
        Asynchronous Clock support
        FrontPanel API Version 12
    Host Platform:
        Compiler: GCC 8.3.0                                                    46:48
        Memory Access: Little Endian: ARM (Release Build) on Nov  2 2019 at 11:4
        Memory Pointer Size: 32 bits
        Large File (>2GB) support
        SDL Video support: No Video Support
        RegEx support for EXPECT commands
        OS clock resolution: 1ms
        Time taken by msleep(1): 1ms                                           9 armv7l GNU/Linux
        git commit id: 4e0450cf.19.75-v7+ #1270 SMP Tue Sep 24 18:45:11 BST 2019
        git commit time: 2019-04-18T20:03:17-07:00
```

- #### the expected behavior and the actual behavior

Expected behaviour is that `SING_STEP` single-steps through the main phases of the instruction (fetch, execute, ...) as opposed to `SING_INST` which simply executed the instruction as a whole. 

Current behaviour is that `SING_STEP` seems to behave like `CONT` – no stepping at all. 

- #### I’m willing to fix this

I can code. At least I think I have some moderate skills. I am happy to try implementing this, but I would like to discuss this with some more experienced expert first.

**Comment by markpizz** [comment]
Where did you come to believe that SING_STEP and SING_INST were commands supported by the PDP8 simulator?

These are not part of the core simh PDP8 simulator.

The simh PDP8 simulator models instruction execution and not processor cycles that may have been how the original hardware may have worked.

You're not running a vanilla simh PDP8, but maybe you're running the PiDP8 variant.  I'm pretty sure that the PiDP8 hasn't changed from the base simh PDP8 simulator that much.

**Comment by Thomas-Ganter** [comment]
> Where did you come to believe that SING_STEP and SING_INST were commands supported by the PDP8 simulator?

Well, Single Stepping Instructions clearly is supported (I can `CTRL-E` and then `S` my way commany by command through the coftware). 

Now I was wondering whether the Single Stepping Steps functionality of the original PDP-8 is somewhere to be found. 

> The simh PDP8 simulator models instruction execution and not processor cycles that may have been how the original hardware may have worked.
> You're not running a vanilla simh PDP8, but maybe you're running the PiDP8 variant. I'm pretty sure that the PiDP8 hasn't changed from the base simh PDP8 simulator that much.

I know I use the software from [PiDP-8/I](https://tangentsoft.com/pidp8i) as I got the tarball and built it locally, but the `SIMH` contained therein seems to be plain vanilla simh with all machines except for the `pdp8` stripped away. 

Implementing single stepping to complete the simulation was an extension I felt like should go into the original software. If this is not considered in line with this project's goals, I'll try my luck with the PiDP-8 folks next.

**Comment by markpizz** [comment]
> Well, Single Stepping Instructions clearly is supported (I can CTRL-E and then S my way command by command through the software).

Absolutely, true.  The S you're typing is short for STEP (not SING_STEP), and it takes an optional argument that is the number of "simulator specific cycles" that should be run before the simulator returns from the sim_instr() routine with the status SCPE_STEP.  The simh framework keeps track of the necessary bookkeeping to make this work.  

> I know I use the software from PiDP-8/I as I got the tarball and built it locally, but the SIMH contained therein seems to be plain vanilla simh with all machines except for the pdp8 stripped away.

That is true.

> Implementing single stepping to complete the simulation was an extension I felt like should go into the original software. If this is not considered in line with this project's goals, I'll try my luck with the PiDP-8 folks next.

If you really want to implement this, and it is done in a clean and consistent way in the base simh code, AND it actually works AND it doesn't break anything, we'll be glad to look closely at what you come up with.  The looking closely will probably involve some feedback about details and when everything looks good a final merge.

Notes: 

- What you're proposing will necessitate reorganizing the main instruction loop in the **sim_instr()** routine in pdp8_cpu.c so that each original processor cycle decrements sim_interval and then dispatches to sim_process_events() when sim_interval is <=0 and then returns from **sim_instr()** when the call to sim_process_events() is not SCPE_OK.  When **sim_instr()** starts, it will have to act on which phase/cycle of the instruction execution sequence it had previously completed and perform the next step and when **sim_instr()** exits it will have to save which phase of the instruction cycle that had been completed.
- Most simulators are implemented such that "simulator specific cycles" are actually instructions.  You're proposing to change this base concept for this simulator.  This absolutely implies a reorganization of how the sim_instr() routine works, but it probably also has some subtle implications for at least some of the DEVICEs in this simulator.   The subtle DEVICE implication exists since each of the DEVICEs was implemented after the current choice of cycles == instructions was made.  Changes in most, if not all, of the DEVICEs would likely be limited to changing calls to **sim_activate(uptr, x)** to  **sim_activate(uptr, x*avg_cycles_per_instr)**.  Changes to DEVICEs can wait until after the CPU rework is done and working.
- You should work directly with the simh supplied sources from https://github.com/simh/simh.  The point here is to avoid the PiDP8 modifications which will likely have to be adjusted somewhat after you're done.  Try to fetch, build and run a vanilla simh pdp8 before you start...
- When making changes here, please follow these guidelines: 

1. Source files have CRLF line endings and have tabs expanded to spaces.
2. Since you're changing an existing source modules (as opposed to one you're writing from scratch), please follow the code indentation and brace placement (style issues) already in use in the existing module.

**Comment by Thomas-Ganter** [comment]
Thanks for the pointers @markpizz – started looking into that. 

Had to dig into the maintenance manuals first to see whether this appears doable. The individual processing steps are sufficiently clear documented, so I guess it is codeable. And I feel sufficiently confident that I can do this.  

> If you really want to implement this, and it is done in a clean and consistent way in the base simh code, AND it actually works AND it doesn't break anything, we'll be glad to look closely at what you come up with. The looking closely will probably involve some feedback about details and when everything looks good a final merge.

The cycle == instruction decision and thus not breaking DEVICEs seems to be the biggest hurdle from the above list. But then, I'll either utterly fail and make a complete fool out of myself, or I'll get back with some status and/ or questions in a while. 

8^)

Probably someone should label this as a feature request though…

**Comment by markpizz** [comment]
As you head down this path.  Fundamentally focus on changing the STEP behavior to be cycles rather than instructions.  Once you solve this, if needed, we'll address the subject of instruction stepping.  Depending on how your stuff works out, there may need to be simh framework (SCP) changes to support that.

**Comment by markpizz** [comment]
I had said:
> Depending on how your stuff works out, there may need to be simh framework (SCP) changes to support that.

No additional framework changes will be needed to support this.

You may want to look at Rich Cornwell's PDP10-KA, PDP10-KI and PDP10-KL simulator for an example of how he changes "STEP n" to be cycles vs instructions.
